host_reference/dslash_reference.cpp

```diff
@@ -0,0 +1,413 @@
+      // DOMAIN_WALL END
+    } else {
+      errorQuda("Unsupported dslash_type");
+    }
+
+    if (inv_param.mass_normalization == QUDA_MASS_NORMALIZATION) {
+      ax(0.25 / (kappa5 * kappa5), spinorCheck, V * spinor_site_size * inv_param.Ls, inv_param.cpu_prec);
+    }
+
+  } else if (inv_param.solution_type == QUDA_MATPCDAG_MATPC_SOLUTION) {
+
+    void *spinorTmp = malloc(V * spinor_site_size * host_spinor_data_type_size * inv_param.Ls);
+    ax(0, spinorCheck, V * spinor_site_size, inv_param.cpu_prec);
+
+    // DOMAIN_WALL START
+    if (dslash_type == QUDA_DOMAIN_WALL_DSLASH) {
+      dw_matpc(spinorTmp, gauge, spinorOut, kappa5, inv_param.matpc_type, 0, inv_param.cpu_prec, gauge_param,
+               inv_param.mass);
+      dw_matpc(spinorCheck, gauge, spinorTmp, kappa5, inv_param.matpc_type, 1, inv_param.cpu_prec, gauge_param,
+               inv_param.mass);
+    } else if (dslash_type == QUDA_DOMAIN_WALL_4D_DSLASH) {
+      dw_4d_matpc(spinorTmp, gauge, spinorOut, kappa5, inv_param.matpc_type, 0, inv_param.cpu_prec, gauge_param,
+                  inv_param.mass);
+      dw_4d_matpc(spinorCheck, gauge, spinorTmp, kappa5, inv_param.matpc_type, 1, inv_param.cpu_prec, gauge_param,
+                  inv_param.mass);
+    } else if (dslash_type == QUDA_MOBIUS_DWF_DSLASH) {
+      double _Complex *kappa_b = (double _Complex *)malloc(Lsdim * sizeof(double _Complex));
+      double _Complex *kappa_c = (double _Complex *)malloc(Lsdim * sizeof(double _Complex));
+      for (int xs = 0; xs < Lsdim; xs++) {
+        kappa_b[xs] = 1.0 / (2 * (inv_param.b_5[xs] * (4.0 + inv_param.m5) + 1.0));
+        kappa_c[xs] = 1.0 / (2 * (inv_param.c_5[xs] * (4.0 + inv_param.m5) - 1.0));
+      }
+      mdw_matpc(spinorTmp, gauge, spinorOut, kappa_b, kappa_c, inv_param.matpc_type, 0, inv_param.cpu_prec, gauge_param,
+                inv_param.mass, inv_param.b_5, inv_param.c_5);
+      mdw_matpc(spinorCheck, gauge, spinorTmp, kappa_b, kappa_c, inv_param.matpc_type, 1, inv_param.cpu_prec,
+                gauge_param, inv_param.mass, inv_param.b_5, inv_param.c_5);
+      free(kappa_b);
+      free(kappa_c);
+      // DOMAIN_WALL END
+    } else {
+      errorQuda("Unsupported dslash_type");
+    }
+
+    if (inv_param.mass_normalization == QUDA_MASS_NORMALIZATION) { errorQuda("Mass normalization not implemented"); }
+
+    // free(spinorTmp);
+
+    int vol = inv_param.solution_type == QUDA_MAT_SOLUTION ? V : Vh;
+    mxpy(spinorIn, spinorCheck, vol * spinor_site_size * inv_param.Ls, inv_param.cpu_prec);
+    double nrm2 = norm_2(spinorCheck, vol * spinor_site_size * inv_param.Ls, inv_param.cpu_prec);
+    double src2 = norm_2(spinorIn, vol * spinor_site_size * inv_param.Ls, inv_param.cpu_prec);
+    double l2r = sqrt(nrm2 / src2);
+
+    printfQuda("Residuals: (L2 relative) tol %g, QUDA = %g, host = %g; (heavy-quark) tol %g, QUDA = %g\n",
+               inv_param.tol, inv_param.true_res, l2r, inv_param.tol_hq, inv_param.true_res_hq);
+  }
```

Comment:
This needs the fix from #979 applied to it to check for an unsupported `solution_type`.